File: 2019/rust/day04/src/main.rs
extern crate aoc;

use aoc::aoc_error::AOCError;
use aoc::utils::parse_fields;
use std::io::{self, Read};

type Result<T> = ::std::result::Result<T, Box<dyn ::std::error::Error>>;

fn is_simple_candidate(n: u64) -> bool {
    let mut adjacent_pair = false;

    let digits = aoc::nums::digits(&n.to_string()).unwrap_or(vec![]);
    for (d1, d2) in digits[..digits.len() - 1].iter().zip(digits[1..].iter()) {
        if d2 < d1 {
            return false;
        }
        if d1 == d2 {
            adjacent_pair = true;
        }
    }

    adjacent_pair
}

fn is_candidate(n: u64) -> bool {
    let mut run_length = 1;
    let mut adjacent_pair = false;

    let digits = aoc::nums::digits(&n.to_string()).unwrap_or(vec![]);
    for (d1, d2) in digits[..digits.len() - 1].iter().zip(digits[1..].iter()) {
        if d2 < d1 {
            return false;
        } else if d1 == d2 {
            run_length += 1;
        } else {
            if run_length == 2 {
                adjacent_pair = true;
            }
            run_length = 1;
        }
    }

    adjacent_pair || run_length == 2
}

fn part1(start: u64, end: u64) -> Result<usize> {
    Ok((start..end + 1).filter(|&n| is_simple_candidate(n)).count())
}

fn part2(start: u64, end: u64) -> Result<usize> {
    Ok((start..end + 1).filter(|&n| is_candidate(n)).count())
}

fn main() -> Result<()> {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input)?;

    let nums = parse_fields(input.trim(), '-')?;
    if nums.len() != 2 {
        Err(AOCError::new("Invalid input"))?;
    }
    let start = nums[0];
    let end = nums[1];

    println!("Part 1: {}", part1(start, end)?);
    println!("Part 2: {}", part2(start, end)?);
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_is_simple_candidate() {
        assert!(is_simple_candidate(111111));
        assert!(!is_simple_candidate(23450));
        assert!(!is_simple_candidate(123789));
    }

    #[test]
    fn test_is_candidate() {
        assert!(is_candidate(112233));
        assert!(!is_candidate(123444));
        assert!(is_candidate(111122));
    }
}

File: 2019/golang/day04/main.go
package main

import (
	"fmt"
	"julianandrews/adventofcode/aoc"
	"os"
	"strconv"
	"strings"
)

func getDigits(n int) []int {
	var digits []int
	for n > 0 {
		digits = append(digits, n%10)
		n /= 10
	}

	// reverse the digits
	for i := len(digits)/2 - 1; i >= 0; i-- {
		opp := len(digits) - 1 - i
		digits[i], digits[opp] = digits[opp], digits[i]
	}

	return digits
}

func nonDecreasing(digits []int) bool {
	for i, d := range digits[1:] {
		if d < digits[i] {
			return false
		}
	}
	return true
}

func hasAdjacentPair(digits []int) bool {
	for i, d := range digits[1:] {
		if d == digits[i] {
			return true
		}
	}
	return false
}

func hasExactRun(n int, digits []int) bool {
	runLength := 1
	for i, d := range digits[1:] {
		if d == digits[i] {
			runLength++
		} else {
			if runLength == n {
				break
			}
			runLength = 1
		}
	}
	return runLength == n
}

func isSimpleCandidate(n int) bool {
	digits := getDigits(n)
	return len(digits) == 6 && nonDecreasing(digits) && hasAdjacentPair(digits)
}

func isCandidate(n int) bool {
	digits := getDigits(n)
	return len(digits) == 6 && nonDecreasing(digits) && hasExactRun(2, digits)
}

func part1(start int, end int) uint {
	count := uint(0)
	for x := start; x <= end; x++ {
		if isSimpleCandidate(x) {
			count++
		}
	}
	return count
}

func part2(start int, end int) uint {
	count := uint(0)
	for x := start; x <= end; x++ {
		if isCandidate(x) {
			count++
		}
	}
	return count
}

func main() {
	input, err := aoc.GetInput()
	if err != nil {
		fmt.Fprintln(os.Stderr, "Failed to read input")
		os.Exit(1)
	}
	parts := strings.Split(strings.TrimSpace(input), "-")
	if len(parts) != 2 {
		fmt.Fprintln(os.Stderr, "Invalid range")
	}
	start, err := strconv.Atoi(parts[0])
	if err != nil {
		fmt.Fprintln(os.Stderr, "Failed to parse start")
		os.Exit(1)
	}
	end, err := strconv.Atoi(parts[1])
	if err != nil {
		fmt.Fprintln(os.Stderr, "Failed to parse end")
		os.Exit(1)
	}

	fmt.Println("Part 1: ", part1(start, end))
	fmt.Println("Part 2: ", part2(start, end))
}
